### DISCUSSION TITLE
Automatic Timing Analyzer

### DISCUSSION THREAD

**Discussion by smallketchup82** [ppy/osu#27662] [Ideas] [answered]
It would be nice to have a timing analyzer in the editor which can automatically determine the BPM and offset of an audio file, similar to the existing [timing analyzer](https://osu.ppy.sh/users/126198#:~:text=timing%20analyzer%20v0.32.4) that has existed for over a decade. This would be a good way for mappers to get timing information for simple songs, or a starting point for timing more complex songs. It would speed up the mapping workflow overall.

It shouldn't be hard to achieve high accuracy (possibly sub-ms) with newer [bpm detection algorithms](https://www.parallelcube.com/web/wp-content/uploads/2018/03/BeatDetectionAlgorithms.pdf). To find offset, all that really needs to be done is to find the first peak (or sound) in the song.

Useful documentation:
- https://stackoverflow.com/questions/657073/how-to-detect-the-bpm-of-a-song-in-php (php but still a good thread due to the discussion of algorithms)
- https://stackoverflow.com/questions/2781897/programmatically-get-bpm-of-a-wave-or-mp3-from-net
- http://www.rotorbrain.com/foote/papers/icme2001/icmehtml.htm
- https://www.parallelcube.com/web/wp-content/uploads/2018/03/BeatDetectionAlgorithms.pdf
- http://archive.gamedev.net/archive/reference/programming/features/beatdetection/index.html
- https://stackoverflow.com/questions/3260/peak-detection-of-measured-signal/

**Comment by peppy** [comment]
Timing is part of the fun of mapping.

There's plenty of external tools for this if a mapper would prefer to bypass that fun.

**Reply by dawidseipold** [reply]
can't see any feasible reality where looking for the offset is fun...

**Comment by c-sig** [comment]
Aside from timing being essential to the mapping experience, additions in the new editor (mainly waveforms) make it insanely easy to time anyway.

**Comment by Ceynou** [comment]
This [paper](https://github.com/nathanstep55/bpm-offset-detector/blob/main/original-paper/report.pdf) is more recent and the result is more accurate, I see absolutely nothing fun in timing something, and as a player, I see absolutely nothing fun playing badly timed songs, nobody times it consistently wrong so you can correct it for every song.

**Reply by Rekunan** [reply, marked as answer]
You bring up a good point that it being fun is very anecdotal. I think a bigger problem would be the work required to create, let alone maintain, such a tool. Considering external tools already exist, I think it's better to just focus that work on more pressing issues.

I also understand that for most ppl, poorly timed maps are unfun, so I would recommend sticking to ranked maps, that aren't too old, to avoid them. If you can spare the time, maybe message the mapper to let them know about it, or even try fixing it yourself.

**Reply by smallketchup82** [reply]
Technically, it could be done [using BASS](https://github.com/peppy/ManagedBass/blob/fork/src/AddOns/BassFx/BassFx.cs) and wouldn't be too hard to implement. I could see it being implemented in 100-200 lines. It's really just an issue of "we have better things we could be working on"
